hail/src/main/scala/is/hail/expr/types/physical/PSubsetStruct.scala

```diff
@@ -0,0 +1,215 @@
+package is.hail.expr.types.physical
+
+import is.hail.annotations.{Region, UnsafeUtils}
+import is.hail.asm4s.{Code, Settable, SettableBuilder, Value, const}
+import is.hail.expr.ir.{EmitCodeBuilder, EmitMethodBuilder, IEmitCode}
+import is.hail.expr.types.BaseStruct
+import is.hail.expr.types.virtual.TStruct
+import is.hail.utils._
+
+object PSubsetStruct {
+  def apply(ps: PStruct, fieldNames: String*): PSubsetStruct = {
+    val f = fieldNames.toArray
+    PSubsetStruct(ps, f)
+  }
+}
+
+// Semantics: PSubsetStruct is a non-constructible view of another PStruct, which is not allowed to mutate
+// that underlying PStruct's region data
+final case class PSubsetStruct(ps: PStruct, _fieldNames: Array[String]) extends PStruct {
+  override lazy val fieldNames = _fieldNames
+  val fields = fieldNames.map(f => ps.field(f)).toFastIndexedSeq
+  val required = ps.required
+
+  if (fields == ps.fields) {
+    log.warn("PSubsetStruct used without subsetting input PStruct")
+  }
+
+  private val idxMap: Array[Int] = fields.zipWithIndex.map { case (f, i) =>
+    val psField = ps.field(f.name)
+    assert(f == psField)
+    psField.index
+  }.toArray
+
+  override lazy val fieldIdx: Map[String, Int] = fields.zipWithIndex.map { case (f, i) =>
+    assert(f.index >= i)
+    (f.name, i)
+  }.toMap
+
+  lazy val missingIdx: Array[Int] = idxMap.map(i => ps.missingIdx(i))
+  lazy val nMissing: Int = missingIdx.length
+
+  override lazy val virtualType = TStruct(fields.map(f => (f.name -> f.typ.virtualType)):_*)
+  override val types: Array[PType] = fields.map(_.typ).toArray
+
+  // FIXME: should structFundamentalType be a ps.selectFields(fieldNames).structFundamentalType?
+  lazy val structFundamentalType: PStruct = ps.structFundamentalType
```

Comment:
I think this should be a PSubsetStruct(ps.structFundamenetalType, _fieldNames)

Reply:
What does fundamental mean here? Is it constructible (because I think right now that's how we're treating it)?